usr/src/cmd/fs.d/smbclnt/fksmbcl/fksmbcl_main.c

```diff
@@ -0,0 +1,939 @@
+	mdata.devfd = ctx->ct_dev_fd;
+
+	/* Build mount args */
+	bzero(&ma, sizeof (ma));
+	ma.spec = mnt_resource;
+	ma.dir = "/";
+	ma.flags =  MS_DATA | MS_OPTIONSTR | MS_NOSPLICE | MS_NOSUID;
+	ma.fstype = "smbfs";
+	ma.dataptr = (void *) &mdata;
+	ma.datalen = sizeof (mdata);
+	ma.optptr = mnt_opt_buf;
+	ma.optlen = sizeof (mnt_opt_buf);
+
+	error = fake_domount("smbfs", &ma, &vfsp);
+	if (error != 0) {
+		fprintf(stderr, "domount error=%d\n", error);
+	}
+
+	/* Mount takes a ref, so always rele here. */
+	(void) nsmb_ioctl(ctx->ct_dev_fd, SMBIOC_TREE_RELE, NULL);
+}
+
+void
+do_unmount(int argc, char **argv)
+{
+	int error;
+
+	if (vfsp == NULL) {
+		fprintf(stderr, "Not mounted\n");
+		return;
+	}
+
+	error = fake_dounmount(vfsp, 0);
+	if (error != 0) {
+		fprintf(stderr, "dounmount error=%d\n", error);
+		return;
+	}
+	vfsp = NULL;
+}
+
+void
+do_statfs(int argc, char **argv)
+{
+	statvfs64_t st;
+	int error;
+
+	if (vfsp == NULL) {
+		fprintf(stderr, "Not mounted\n");
+		return;
+	}
+
+	error = fsop_statfs(vfsp, &st);
+	if (error != 0) {
+		fprintf(stderr, "df error=%d\n", error);
+		return;
+	}
+	printf("bsize=%ld\n", st.f_bsize);
+	printf("frsize=%ld\n", st.f_frsize);
+	printf("blocks=%" PRIu64 "\n", st.f_blocks);
+	printf(" bfree=%" PRIu64 "\n", st.f_bfree);
+	printf("bavail=%" PRIu64 "\n", st.f_bavail);
+}
+
+void
+do_dir(int argc, char **argv)
+{
+	char *rdir;
+	vnode_t *vp = NULL;
+	offset_t off;
+	int cnt;
+	int error;
+
+	if (vfsp == NULL) {
+		fprintf(stderr, "mnt required first\n");
+		return;
+	}
+	if (argc > 1)
+		rdir = argv[1];
+	else
+		rdir = "";
+
+	error = vn_open(rdir, 0, FREAD, 0, &vp, 0, 0);
+	if (error != 0) {
+		fprintf(stderr, "do_dir, vn_open error=%d\n", error);
+		return;
+	}
+
+	off = 0;
+	do {
+		cnt = fake_getdents(vp, &off, tbuf, TBUFSZ);
+		if (cnt < 0) {
+			fprintf(stderr, "do_dir, getdents %d\n", -cnt);
+			break;
+		}
+		show_dents(vp, &off, tbuf, cnt);
+	} while (cnt > 0);
+
+	if (vp != NULL)
+		vn_close_rele(vp, 0);
+}
+
+void
+do_dirx(int argc, char **argv)
+{
+	char *rdir;
+	vnode_t *vp = NULL;
+	offset_t off;
+	int cnt;
+	int error;
+
+	if (vfsp == NULL) {
+		fprintf(stderr, "mnt required first\n");
+		return;
+	}
+	if (argc > 1)
+		rdir = argv[1];
+	else
+		rdir = "";
+
+	error = vn_open(rdir, 0, FREAD|FXATTRDIROPEN, 0, &vp, 0, 0);
+	if (error != 0) {
+		fprintf(stderr, "do_dirx, vn_open error=%d\n", error);
+		return;
+	}
+
+	off = 0;
+	do {
+		cnt = fake_getdents(vp, &off, tbuf, TBUFSZ);
+		if (cnt < 0) {
+			fprintf(stderr, "do_dirx, getdents %d\n", -cnt);
+			break;
+		}
+		show_dents(vp, &off, tbuf, cnt);
+	} while (cnt > 0);
+
+	if (vp != NULL)
+		vn_close_rele(vp, 0);
+}
+
+static void
+show_dents(vnode_t *dvp, offset_t *offp, char *buf, int cnt)
+{
+	char time_buf[40];
+	struct stat64 st;
+	vnode_t *vp;
+	char *p;
+	dirent_t *d;
+	offset_t offset = (offset_t)-1L;
+	int error;
+	uint_t mode;
+	char type;
+
+	p = buf;
+	while (p < (buf + cnt)) {
+		/* LINTED E_BAD_PTR_CAST_ALIGN */
```

Comment:
So this is another case where  'void *' is handy.  Arguably p in this case really is 'void *', but the problem is that you can't do arithmetic on untyped pointers.  I'd just stick a (void *) cast here.

Reply:
OK